### PR TITLE
Fix performance issues with many offline sessions

### PR DESCRIPTION
**Pull request by hmlnarik**
Fixes: #13340

Pipeline run: https://master-jenkins.redhat.com/job/universal-test-pipeline-server/2385/

### CODE REVIEW

File: model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java [hunk 1]
```diff
@@ -379,10 +379,19 @@ public Stream<UserSessionModel> loadUserSessionsStream(Integer firstResult, Inte
...
     }
 
-    private Stream<UserSessionModel> loadUserSessionsWithClientSessions(TypedQuery<PersistentUserSessionEntity> query, String offlineStr) {
+    /**
+     *
+     * @param query
+     * @param offlineStr
+     * @param useExact If {@code true}, then only client sessions from the user sessions
+     *          obtained from the {@code query} are loaded. If {@code false}, then user session
+     *          is taken as range, and all client sessions in between the minimum and maximum session
```
Comment (sguilhen) [inline model/jpa/src/main/java/org/keycloak/models/jpa/session/JpaUserSessionPersisterProvider.java:391]: minor issue: the sentence doesn't look finished, seems like " are loaded" is missing here.

**Review by sguilhen — COMMENTED** [review]
LGTM, thanks @hmlnarik . Only one comment that doesn't really prevent merging this.

**Review by sguilhen — APPROVED** [review]

**Comment by martin-kanis** [comment]
Thanks for the PR. I have been thinking about this a bit and here are my thoughts: 
Should we be worried about reintroducing IN-clause which was removed in last performance optimization [attempt](https://github.com/keycloak/keycloak/pull/8012)? If the `userSessionsIds` set used in IN-cause gets too big, performance will be probably bad. However, I think we should be OK because `userSessionIds` size should be reasonable. There are these use-cases: 
- load user session by user (not limited by first,max parameters as it is called with first=0, max=null, but limited by number of user session which user can has. Not sure if this can be a high number in real scenarios. Maybe yes, if someone creates many offline user sessions for a user by i.e `password` grant type)
- load user sessions by client (this can be a high number but it is called only from services and endpoint is limiting max result size -- 100 by default)
- load user sessions by interval - IN clause not used and method is called only from pre-loading

@hmlnarik What are your thoughts on this?

**Comment by hmlnarik** [comment]
> @hmlnarik What are your thoughts on this?

Thank you for the reply. These are exactly the arguments why I am in favour of introducing the `IN` clauses for the cases when the session is known in advance while keeping the interval for preloading, and this PR should reflect it.

**Review by martin-kanis — APPROVED** [review]

**Review by mhajas — APPROVED** [review]
Thanks for the PR!

**Pull request merged**